proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,180 @@
+interface.  The aforementioned function signature that most sysvar getters use
+is also inefficient for large accounts.  Copying the entire sysvar to VM memory
+might consume excessive compute units (e.g. 16392 bytes for 'stake history').
+
+### Unblocking core BPF programs
+
+As per [SIMD-0088](./0088-enable-core-bpf-programs.md), an effort is underway
+to reduce code complexity of the Solana runtime by porting native programs to
+core BPF programs.
+
+These new programs target full compatibility with their predecessors to avoid
+introducing breaking API changes.  In some cases, native programs read a sysvar
+from the sysvar cache without requiring the user to include this sysvar in the
+transaction account list.  This happens in the stake program 'redelegate'
+instruction, for example.
+
+As a result, it is currently impossible to port certain native programs to core
+BPF programs.  Another factility to expose sysvars to the virtual machine is
+required.
+
+### Optimizing existing user programs
+
+Reading sysvars with the proposed syscall is expected to be more efficient than
+an explicit account access.  Removing explicit account accesses to sysvars will
+generally result in smaller transactions.  Reducing the amount of account data
+that is mapped into virtual machine memory may also yield cost savings.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-specific syscalls
+
+TODO
+
+### New memory regions
+
+TODO
+
+## Detailed Design
+
+### Sysvar Cache
+
+TODO
+
+### Restrictions on sysvar data dependencies
+
+TODO
+
+### `sol_get_sysvar` syscalls
+
+TODO this is outdated
+
+```c
+/* sol_get_sysvar_size queries the serialized byte size of a sysvar.
+   It is equal to the length of the serialization of the sysvar value.
+   The sysvar value is taken in the sysvar cache.
+
+   sysvar_id points to the 32 byte address of the sysvar.
+
+   If a non-empty sysvar was found in the sysvar cache, returns the
+   byte size of that sysvar.
+
+   If the requested sysvar was not found, is corrupt, is
+   unsupported for the current feature set, or is zero size, returns 0.
+
+   This syscall aborts the transaction if (and only if):
+   - The compute budget was exhausted */
+
+uint64_t
+sol_get_sysvar_size( /* r1 */ void const * sysvar_id );
+
+/* sol_get_sysvar copies sysvar data to VM memory.
+
+   sysvar_id points to the 32 byte address of the sysvar.
+
+   A sysvar is valid if sol_get_sysvar_size(sysvar_id) returns non-zero.
+   If the sysvar is not valid, returns zero without writing to out.
+
+   The byte range [offset,offset+size) selects the range of data to be
+   copied. The sysvar data is the serialized value of the corresponding
+   sysvar cache entry.
+   The size of this sysvar data is returned by sol_get_sysvar_size.
+
+   Copies the sysvar data to [out,out+size) and returns 1 if the sysvar is valid.
```

Comment:
It sounds like this does offer a lot of flexibility for larger sysvars, particularly when only a subset or a single element of the data is being requested. However, this still would not enable a BPF program to request the full data for a sysvar like `SlotHashes`, correct? The program would still face the current "too large to copy" resource limitation?

Reply:
if you really wanted to, you could take the whole thing to heap. i believe you said slot hashes is like 18kb? do you need more than one entry for your usecase tho? i only need one for mine

Reply:
ALT nope but Vote does yeah. Unfortunately, Vote needs something like `SlotHashes::get_slots() -> &[Slot]`.

Reply:
Aside from that, it seems like single-element lookups would suffice for all the other use cases I have. There's three I can think of across ALT, Vote, and Feature Gate.